BREAKING CHANGE: use definitions data from w3c/webref

dontcallmedom opened the pull request:
close #69
Compared to the current bikeshed data, this only attemps to load data from editors draft.

Adding support for loading data from TR wouldn't be particularly hard, but would definitely bloat the definition files; since I'm not clear the distinction between "current" and "snapshot" is really used in xref today, I thought I would start with the most useful data set.

dontcallmedom committed:
Obtain definitions data from webref instead of bikeshed

close #69
Compared to the current bikeshed data, this only attemps to load data from editors draft.

sidvishnoi commented:
@dontcallmedom Thanks for starting work on this! I wanted to get to it sooner.
I think this would require a careful rewrite for scraper and search, as we don't enough tests.

sidvishnoi commented on scraper.ts:
```diff
@@ -163,32 +149,18 @@ function parseData(content: string, errorURIs: string[], trie: Trie) {
-    for: dataFor.length > 0 ? dataFor : undefined,
+    shortname: series,
+    status: "current",
+    uri: '#' + dfn.id, // This is full URL to term here
```
This is kinda breaking change, for HTML specs, as it no longer has the path.

```diff
-   "uri": "iframe-embed-object.html#attr-object-type",
+   "uri": "#attr-object-type",
```

sidvishnoi replied:
Instead of `dfn.id`, we should be using `dfn.href.replace(spec.url)`. I want to avoid the multipage redirect.

https://html.spec.whatwg.org/multipage/#attr-object-type -> https://html.spec.whatwg.org/multipage/iframe-embed-object.html#attr-object-type

dontcallmedom committed:
Fix handling of definitions in multipage specs

Per https://github.com/sidvishnoi/respec-xref-route/pull/70#discussion_r536075817

sidvishnoi committed:
add webref type dfns to separate file

sidvishnoi committed:
remove unused errorURIs

sidvishnoi committed:
use types from 'webref'

sidvishnoi committed:
nit: extract readJSON

sidvishnoi committed:
nit: combined mapDefinition param into specMetaData

sidvishnoi commented on scraper.ts:
```diff
@@ -223,38 +169,43 @@ function normalizeTerm(term: string, type: string) {
 
   const specMap: Store['specmap'] = Object.create(null);
   const specUrls = new Set<string>();
+  const dfnSources = new Set<DfnsJSON>();
```
This should probably be an array.

sidvishnoi commented on scraper.ts:
```diff
@@ -223,38 +169,43 @@ function normalizeTerm(term: string, type: string) {
+
+  for (const entry of data) {
+    specUrls.add(entry.nightly.url);
+    if (entry.release && entry.release.url) specUrls.add(entry.release.url);
```
```suggestion
    if (entry.release?.url) specUrls.add(entry.release.url);
```

dontcallmedom committed:
Use simpler syntax on optional entries

Co-authored-by: Sid Vishnoi <sidvishnoi8@gmail.com>

sidvishnoi committed:
run prettier

sidvishnoi commented on scraper.ts:
```diff
@@ -1,21 +1,20 @@
 import { promises as fs, existsSync } from 'fs';
 import { resolve as resolvePath, join as joinPath } from 'path';
 import { spawn } from 'child_process';
-import Trie from 'compact-prefix-tree/cjs';
```
Lets also `npm un compact-prefix-tree --save` and `rm @types/compact-prefix-tree.d.ts`

sidvishnoi replied:
We're doing pair programming here :smile: 
Lets do live coding someday!

sidvishnoi commented on scraper.ts:
```diff
@@ -1,21 +1,20 @@
-// Reads and parses anchor data files from bikeshed-data repository and writes:
+// Reads and parses anchor data files from webref repository and writes:
```
```suggestion
// Reads and parses definition data files from webref repository and writes:
```

dontcallmedom committed:
Fix comment

Co-authored-by: Sid Vishnoi <sidvishnoi8@gmail.com>

sidvishnoi commented on scraper.ts:
```diff
@@ -223,38 +171,43 @@ function normalizeTerm(term: string, type: string) {
-  const urlFileContent = await readFile(SPECS_JSON, 'utf8');
-  const data: SpecsJSON = JSON.parse(urlFileContent);
+  const urlFileContent = await readJSON(SPECS_JSON);
+  const data: Array<SpecsJSON> = urlFileContent.results;
```
For consistency..

```suggestion
  const data: SpecsJSON[] = urlFileContent.results;
```

sidvishnoi commented on scraper.ts:
```diff
@@ -223,38 +171,43 @@ function normalizeTerm(term: string, type: string) {
     };
   }
 
   const urls = [...specUrls].sort();
```
Looks like we don't need it anymore.

```suggestion
```

dontcallmedom committed:
Simplify array typing

Co-authored-by: Sid Vishnoi <sidvishnoi8@gmail.com>

dontcallmedom committed:
Remove dead code

Co-authored-by: Sid Vishnoi <sidvishnoi8@gmail.com>

sidvishnoi commented on scraper.ts:
```diff
@@ -223,38 +171,42 @@ function normalizeTerm(term: string, type: string) {
-    specMap[spec] = {
-      url: entry.current_url || entry.snapshot_url,
+    specMap[entry.shortname] = {
+      url: entry.nightly.url || (entry.release ? entry.release.url : entry.url),
```
```suggestion
      url: entry.nightly.url || entry.release?.url || entry.url,
```

sidvishnoi committed:
remove dead-er code

sidvishnoi committed:
use array for dfnSources

sidvishnoi committed:
rename getSpecsData to getAllData

dontcallmedom committed:
Use simpler syntax on optional entries

Co-authored-by: Sid Vishnoi <sidvishnoi8@gmail.com>

sidvishnoi commented:
I'm testing this on some real specs locally..

sidvishnoi commented on scraper.ts:
```diff
@@ -163,32 +119,24 @@ function parseData(content: string, errorURIs: string[], trie: Trie) {
-    uri, // This is full URL to term here
-    normative: normative === '1',
-    for: dataFor.length > 0 ? dataFor : undefined,
+    spec: spec.spec,
+    shortname: spec.shortname,
```
Found the ~bug~ regression :tada: 
These need to be lowercase.

sidvishnoi commented on scraper.ts:
```diff
@@ -223,38 +171,42 @@ function normalizeTerm(term: string, type: string) {
   return term;
```
This helps with another regression in search..
```suggestion
  if (type === "dfn") {
    return term.toLowerCase();
  }
  return term;
```

sidvishnoi commented on scraper.ts:
```diff
@@ -223,38 +171,42 @@ function normalizeTerm(term: string, type: string) {
+    specMap[entry.shortname] = {
+      url: entry.nightly.url || entry.release?.url || entry.url,
       title: entry.title,
       shortname: entry.shortname,
```
```suggestion
      shortname: entry.series.shortname.toLowerCase(),
```

sidvishnoi commented on scraper.ts:
```diff
@@ -223,38 +171,42 @@ function normalizeTerm(term: string, type: string) {
-
-    specMap[spec] = {
-      url: entry.current_url || entry.snapshot_url,
+    specMap[entry.shortname] = {
```
```suggestion
    specMap[entry.shortname.toLowerCase()]
```

sidvishnoi commented:
After the fixes in review above:


## No change / All good 

audiobooks, charmod-norm, did-core, did-spec-registries, fingerprinting-guidance, gamepad, json-ld-syntax, mediacapture-screen-share, navigation-timing, page-visibility, payment-request, pub-manifest, screen-wake-lock, touch-events, tt-profile-registry, web-share, webrtc

## Gone Bad

None.

## Kinda Broken

- manifest: one breakage (looks good, it linked to `link type "manifest"` which is a heading but bikeshed-data/HTML had it as "dfn": https://github.com/tabatkins/bikeshed-data/blob/dec016a9c2f9b9b32dd1e4c286a42f95ae5d5e8f/data/anchors/anchors-li.data#L3446-L3454)
- badging: `link type "icon"`, same as above
- web-nfc: due to addition to "records" from web-nfc itself, "record" (as in `webidl#idl-record`) in pluralized form (records) is not found. this is more of a usage/search algorithm issue.
- 2 test cases in respec, due to updated/improved data

sidvishnoi committed:
Merge branch 'master' into reffy-src

sidvishnoi commented on scraper.ts:
```diff
@@ -148,32 +104,24 @@ function parseData(content: string, errorURIs: string[], trie: Trie) {
-    normative: normative === '1',
-    for: dataFor.length > 0 ? dataFor : undefined,
+    spec: spec.spec,
+    shortname: spec.shortname,
```
```suggestion
    shortname: spec.shortname.toLowerCase(),
```

sidvishnoi commented on scraper.ts:
```diff
@@ -148,32 +104,24 @@ function parseData(content: string, errorURIs: string[], trie: Trie) {
-    uri, // This is full URL to term here
-    normative: normative === '1',
-    for: dataFor.length > 0 ? dataFor : undefined,
+    spec: spec.spec,
```
```suggestion
    spec: spec.spec.toLowerCase(),
```

sidvishnoi committed:
fix regressions

sidvishnoi commented on scraper.ts:
```diff
@@ -205,41 +153,48 @@ function normalizeTerm(term: string, type: string) {
-
-    specMap[spec] = {
-      url: entry.current_url || entry.snapshot_url,
+    specMap[entry.shortname.toLowerCase()]
```
```suggestion
    specMap[entry.shortname.toLowerCase()] = {
```

sidvishnoi committed:
fix a my bad

sidvishnoi committed:
remove unused compact-prefix-tree dependency

sidvishnoi reviewed (approved):
:shipit:
